I2C Bringup

ethanmpeterson opened the pull request.

ologush committed:
Started I2C bringup

ethanmpeterson committed:
Base RTC initialization and loading a default timestamp (#1)

- Load default timestamp in main.c and display it on UART
- Add UART receive hooks.
- Still need UART receive interrupts to properly handle an incoming
  timestamp string

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+
+    //Configure timings
+    //Generated with CubeMX might not be best practice
+    I2C1->TIMINGR = 0x00B07CB4;
```
Interesting, CUBEMX gave you this magic number? Also see this register has a bunch of fields, might be more readable to decompose it into the fields and comment on each one you're setting.

ologush replied:
Yes that's a great Idea, I will break it down and comment on each location. The manual suggested using CubeMX to generate the value to store in the register.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+}
+
+//This works with 7-bit address. For a 10 bit address either make a seperate function or handle both here
+hal_error_E hal_i2c_setAddr(uint8_t addr) {
```
My vote would be to use the mode setting to throw an error if the address exceeds a certain width. You can use a uint16_t in all cases that way.

ethanmpeterson replied:
Another thought, should this be it's own function at all? Or should we pass the address as an arg to a send and receive function?

ologush replied:
I had based this off of the wire arduino library. However, for our case I agree, it makes more sense to pass it all through as one funciton.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.h:
```diff
@@ -0,0 +1,33 @@
+typedef enum {
+    HAL_I2C_ADDRESS_MODE_7_BIT,
+    HAL_I2C_ADDRESS_MODE_10_BIT
+} hal_i2c_address_mode;
```
Nit on enum type naming: It does not really matter but in the rest of the repo I use _E for enums and _S for structs as a suffix. Would be good to stick to that here if possible. Naming of the enums fields makes sense though.

ologush replied:
Makes sense, I will follow this formatting. I am now realizing this enum will likely not be necessary as once I2C is set up you can switch between address modes. I will leave this for now until we decide on whether to include it.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+#include "stm32l0xx.h"
+#include "system_stm32l0xx.h"
+
+#include "shared_defs.h"
```
Might be able to remove this import if you are not using the macros within.

ologush replied:
Makes sense, will add later if I need one of the macros, or create one that can be used across other files.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.h:
```diff
@@ -0,0 +1,33 @@
+extern "C" {
+#endif
+
+typedef enum {
```
These typedefs can be shared between different microcontrollers, I think it is fair to say many others will support the same basic functionality shown here. I think these should live in a `hal_i2c.h` in the hal repo and not be microspecific.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+    //This is to configure it in master mode
+
+    //clear PE bit
+    I2C1->CR1 &= ~(1<<I2C_CR1_PE_Pos);
```
Nit: spaces around left shift operator?

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+    I2C1->CR1 |= 1 << I2C_CR1_PE_Pos;
+}
+
+hal_error_E hal_i2c_sendData(uint8_t* data, uint8_t nBytes) {
```
I think we will want to refactor this to be channelized like hal_uart. So these microspecific functions would be made private and passed as pointer to the shared module through a config struct. Then hal_i2c_sendData (and the similar functions defined below) would live in the hal repo, with just the register diddling living here.

This also provides flexibility for the multi i2C bus use case.

ethanmpeterson replied:
We probably also want to think through the master/slave cases and make sure these functions return the correct error codes.

ethanmpeterson commented on src/hal/hal_i2c_microSpecific.c:
```diff
@@ -0,0 +1,77 @@
+    I2C1->CR2 &= nBytes<<I2C_CR2_NBYTES_Pos;
+
+    for (uint8_t i = 0; i < nBytes; i++) {
+        while (!(I2C1->ISR & I2C_ISR_RXNE_Msk));
```
Did a quick read on this and it seems that there is no hardware buffering. So for your initial testing, run this outside of FreeRTOS since the added scheduler latency may lead to dropped bytes. You can also check me by trying it in a freeRTOS task of varying tick rate as well. My guess is that we will need an interrupt to grab bytes and place them in a buffer for us. (but we can cross that bridge later)

ologush committed:
Channelized the i2c hal

ethanmpeterson committed:
Bring up USART Interrupts (#4)

TODO:
- Add ability to register an interrupt handler for bytes received
- Add ability to register an interrupt handler for bytes sent
- Add handler for general peripheral faults

ologush committed:
Fixed compiler issues

ologush commented:
Closed to allow for newest commits to have PR.
